src/generate.rs

```diff
@@ -0,0 +1,211 @@
+use crate::topology::config::{
+    component::ExampleError, GlobalOptions, SinkDescription, SourceDescription,
+    TransformDescription,
+};
+use indexmap::IndexMap;
+use serde::Serialize;
+use std::collections::BTreeMap;
+use structopt::StructOpt;
+use toml::Value;
+
+#[derive(StructOpt, Debug)]
+#[structopt(rename_all = "kebab-case")]
+pub struct Opts {
+    /// Generate expression, e.g. 'stdin|json_parser,add_fields|console'
+    ///
+    /// Three comma-separated lists of sources, transforms and sinks, separated
+    /// by pipes. If subsequent component types are not needed then their pipes
+    /// can be omitted from the expression.
+    ///
+    /// For example:
+    ///
+    /// `|json_parser` prints a `json_parser` transform.
+    ///
+    /// `||file,http` prints a `file` and `http` sink.
+    ///
+    /// `stdin||http` prints a `stdin` source and an `http` sink.
+    ///
+    /// Vector makes a best attempt at constructing a sensible topology. The
+    /// first transform generated will consume from all sources and subsequent
+    /// transforms will consume from their predecessor. All sinks will consume
+    /// from the last transform or, if none are specified, from all sources. It
+    /// is then up to you to restructure the `inputs` of each component to build
+    /// the topology you need.
+    ///
+    /// Generated components are given incremental names (`source1`, `source2`,
+    /// etc) which should be replaced in order to provide better context.
+    expression: String,
+}
+
+#[derive(Serialize)]
+pub struct SinkOuter {
+    pub healthcheck: bool,
+    pub inputs: Vec<String>,
+    #[serde(flatten)]
+    pub inner: Value,
+    pub buffer: crate::buffers::BufferConfig,
+}
+
+#[derive(Serialize)]
+pub struct TransformOuter {
+    pub inputs: Vec<String>,
+    #[serde(flatten)]
+    pub inner: Value,
+}
+
+#[derive(Serialize, Default)]
+pub struct Config {
+    #[serde(flatten)]
+    pub global: GlobalOptions,
+    pub sources: IndexMap<String, Value>,
+    pub transforms: IndexMap<String, TransformOuter>,
+    pub sinks: IndexMap<String, SinkOuter>,
+}
```

Comment:
Yeah it felt a bit off, but it does give us tighter control over what gets generated. I've just updated the sections to be optionals to avoid generating empty sections.